### PR TITLE
Personal invites replacing invite codes

### PR DESCRIPTION
**Pull request by megrogan**

### CODE REVIEW

File: backend/canisters/group/api/can.did [hunk 1]
```diff
@@ -372,7 +317,6 @@ type OptionalGroupPermissions =
         delete_messages: opt PermissionRole;
         update_group: opt PermissionRole;
         pin_messages: opt PermissionRole;
-        invite_users: opt PermissionRole;
```
Comment (hpeebles) [inline backend/canisters/group/api/can.did:375]: Surely the `invite_users` permission is still needed?

File: backend/canisters/group/api/src/queries/public_summary.rs [hunk 2]
```diff
@@ -1,11 +1,9 @@
 use candid::CandidType;
 use serde::{Deserialize, Serialize};
+use types::Empty;
```
Comment (hpeebles) [inline backend/canisters/group/api/src/queries/public_summary.rs:3]: use types::{Empty, PublicGroupSummary};

File: backend/canisters/group/impl/src/model/invited_users.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            self.invites.remove(&principal);
+            self.last_updated = now;
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn get(&self, user_id_or_principal: Principal) -> Option<&UserInvitation> {
+        self.invites
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/model/invited_users.rs:39]: Maybe this?
```
let principal = self.users.get(&user_id_or_principal.into()).unwrap_or(user_id_or_principal);
self.invites.get(&principal)
```

File: backend/canisters/group/impl/src/updates/c2c_invite_users.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    if let Some(participant) = runtime_state.data.participants.get_by_user_id(&args.caller) {
+        // The original caller must be authorized to invite other users
+        if participant.suspended.value
+            || (!runtime_state.data.is_public && !participant.role.can_invite_users(&runtime_state.data.permissions))
+        {
+            return NotAuthorized;
+        }
+
+        // Filter out users who are already members
+        let invited_users: Vec<_> = args
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/updates/c2c_invite_users.rs:37]: Also filter our those who are already invited

File: backend/canisters/group/impl/src/updates/c2c_invite_users.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            .users
+            .iter()
+            .filter(|(_, principal)| runtime_state.data.participants.get_by_principal(principal).is_none())
+            .copied()
+            .collect();
+
+        let user_ids: Vec<_> = invited_users.iter().map(|(user_id, _)| user_id).copied().collect();
+
+        if !runtime_state.data.is_public {
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/updates/c2c_invite_users.rs:46]: Nothing actually happens if the group is public... so surely we can short circuit earlier or reject the call?

File: backend/canisters/group/impl/src/updates/decline_invitation.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    match lookup_user(caller, user_index_canister_id).await {
+        Ok(user) => mutate_state(|state| {
+            let now = state.env.now();
+
+            match state.data.invited_users.remove(&user.user_id, now) {
+                true => Success,
+                false => NotInvited,
+            }
+        }),
+        Err(LookupUserError::UserNotFound) => NotAuthorized,
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/updates/decline_invitation.rs:23]: NotAuthorized is probably unnecessary and we can just return NotInvited in this case.

File: backend/canisters/local_user_index/api/can.did [hunk 7]
```diff
@@ -34,10 +52,14 @@ type ReportMessageArgs =
 type ReportMessageResponse =
     variant {
         Success;
+
```
Comment (hpeebles) [inline backend/canisters/local_user_index/api/can.did:55]: These can go

File: backend/libraries/types/can.did [hunk 8]
```diff
@@ -658,6 +659,7 @@ type Participant =
 type ParticipantJoined =
     record {
         user_id: UserId;
+        invited: bool;
```
Comment (hpeebles) [inline backend/libraries/types/can.did:662]: Maybe `invited_by: Option<UserId>`

File: backend/canisters/group/api/can.did (hunk 1 above)
Comment (megrogan) [reply backend/canisters/group/api/can.did:375]: Yes it got accidentally removed from the candid

**Pull request closed without merging**